cpp/ql/src/semmle/code/cpp/ir/implementation/raw/internal/TranslatedCall.qll

```diff
@@ -336,7 +314,204 @@ class TranslatedStructorCall extends TranslatedFunctionCall {
+        (
+          tse.getCall() = getCall() and
+          tse.getArgumentIndex() = index and
+          if tse.isWrite() then isWrite = 1 else isWrite = 0
+        )
+      |
+        tse order by isWrite, index
+      )
+  }
+
+  override Instruction getChildSuccessor(TranslatedElement te) {
+    exists(int i |
+      getChild(i) = te and
+      if exists(getChild(i + 1))
+      then result = getChild(i + 1).getFirstInstruction()
+      else result = getParent().getChildSuccessor(this)
+    )
+  }
+
+  override predicate hasInstruction(Opcode opcode, InstructionTag tag, Type t, boolean isGLValue) {
+    none()
+  }
+
+  override Instruction getFirstInstruction() { result = getChild(0).getFirstInstruction() }
+
+  override Instruction getInstructionSuccessor(InstructionTag tag, EdgeKind kind) { none() }
+
+  override Instruction getInstructionOperand(InstructionTag tag, OperandTag operandTag) { none() }
+
+  override Type getInstructionOperandType(InstructionTag tag, TypedOperandTag operandTag) { none() }
+
+  /**
+   * Gets the `TranslatedFunction` containing this expression.
+   */
+  final TranslatedFunction getEnclosingFunction() {
+    result = getTranslatedFunction(expr.getEnclosingFunction())
+  }
+
+  /**
+   * Gets the `Function` containing this expression.
+   */
+  override Function getFunction() { result = expr.getEnclosingFunction() }
+}
+
+class TranslatedSideEffect extends TranslatedElement, TTranslatedArgumentSideEffect {
+  Call call;
+
+  Expr arg;
+
+  int index;
+
+  boolean write;
+
+  TranslatedSideEffect() { this = TTranslatedArgumentSideEffect(call, arg, index, write) }
+
+  override Locatable getAST() { result = arg }
+
+  Expr getExpr() { result = arg }
+
+  Call getCall() { result = call }
+
+  int getArgumentIndex() { result = index }
+
+  predicate isWrite() { write = true }
+
+  override string toString() {
+    write = true and
+    result = "(write side effect for " + arg.toString() + ")"
+    or
+    write = false and
+    result = "(read side effect for " + arg.toString() + ")"
+  }
+
+  override TranslatedElement getChild(int n) { none() }
+
+  override Instruction getChildSuccessor(TranslatedElement child) { none() }
+
+  override Instruction getFirstInstruction() { result = getInstruction(OnlyInstructionTag()) }
+
+  override predicate hasInstruction(Opcode opcode, InstructionTag tag, Type t, boolean isGLValue) {
+    isWrite() and
+    hasSpecificWriteSideEffect(opcode) and
+    tag = OnlyInstructionTag() and
+    (
+      index != -1 and
+      t = arg.getUnspecifiedType().(DerivedType).getBaseType()
+      or
+      index = -1 and
+      not arg.getUnspecifiedType() instanceof DerivedType and
+      t = arg.getUnspecifiedType()
+      or
+      index = -1 and
+      t = arg.getUnspecifiedType().(DerivedType).getBaseType()
+    ) and
+    isGLValue = false
+    or
+    not isWrite() and
+    hasSpecificReadSideEffect(opcode) and
+    tag = OnlyInstructionTag() and
+    t instanceof VoidType and
```

Comment:
As @rdmarsh2 and I discussed yesterday, it seems limiting to put `VoidType` here. If this instruction conceptually just dereferences its memory and then discards the value, then what can we use it for? If it's just a place to put the memory operand, maybe it would be simpler to put it on the call itself.

On the other hand, I'm not sure if it's helpful to make it return, say, an `int` when the corresponding argument is `int *`. What will the `InitializeParameterPointee` (or whatever) instruction on the caller side look like? I imagine it'll have a _memory_ result and no operands. So for data flow we'd want to connect the _memory_ edges from the caller to the callee, and that doesn't involve the return value of this instruction. Maybe this instruction should return the memory operand that it took as input?

A related question is: should a must-write side effect have a memory input? What does it conceptually use that input for?

Reply:
1. I don't think we'd use the result of a read side effect instruction for anything. It's morally equivalent to an argument operand to a call. The reason we don't put it on the call itself it because then we need a system for associating the address operand of the read side effect with the memory operand of the read side effect, the index of the argument being dereferenced, etc., all of which we can already do by making it a separate instruction. Other compilers I've worked on have had more complicated operands (e.g. constants, complex addressing modes, etc.) in order to avoid having to create multiple instructions for an operation that will likely boil down to a single machine instruction later. Since we don't care about that particular concern, having more instructions but simple operands seemed like the right way to go.

2. `InitializeParameterPointee` (or whatever) is morally equivalent to `InitializeParameter`. It'll take an `AddressOperand` with the value of the parameter, and return a memory result representing the initialized memory.

3. Although I was the one who originally specified the forms of the various side-effect instructions, I now believe that all write side effects should have a memory result, an `AddressOperand` (except for `CallWriteSideEffect`), a `BufferSizeOperand` (only for buffer write side effects), and **no** memory operand. We'll insert a `Chi` during SSA construction if necessary.

Reply:
I'm with @dave-bartolomeo on 1 and 2. I need to think about 3 - I'm not sure that the semantics we currently give to `Chi` nodes line up with the semantics that the merge node for a `MayWrite` would need.

Reply:
I think what's making you uneasy is that we don't currently differentiate between a may-write and a must-write in SSA. Right now, we treat everything as a must-write. To handle a may-write, all we have to do is treat any overlap result other than no overlap as `MayPartiallyOverlap`, and `Chi` nodes should magically work correctly.